tiledb/sm/buffer/buffer.h

```diff
@@ -44,18 +43,91 @@ namespace sm {
 
 class ConstBuffer;
 
-/** Enables reading from and writing to a buffer. */
-class Buffer {
+// =======================================================
+//      BufferBase
+// =======================================================
+/**
+ * Base class for `Buffer` and `ConstBuffer`
+ *
+ * Responsible for maintaining an read offset in the range [0..size_]. Not
+ * responsible for memory management.
+ */
+class BufferBase {
  public:
-  /* ********************************* */
-  /*     CONSTRUCTORS & DESTRUCTORS    */
-  /* ********************************* */
+  /** Returns the buffer size. */
+  uint64_t size() const;
+
+  /** Returns the current read position, in bytes. */
+  uint64_t offset() const;
+
+  /** Resets the buffer offset to 0. */
+  void reset_offset();
+
+  /** Sets the buffer offset to the input offset. */
+  void set_offset(uint64_t offset);
+
+  /** Advances the offset by *nbytes*. */
+  void advance_offset(uint64_t nbytes);
+
+  /** Checks if reading has reached the end of the buffer. */
+  bool end() const;
+
+  /**
+   * Reads from the local data into the input buffer.
+   *
+   * @param destination The buffer to read the data into.
+   * @param nbytes The number of bytes to read.
+   * @return Status
+   */
+  Status read(void* destination, uint64_t nbytes);
+
+ protected:
+  BufferBase();
+  BufferBase(void* data, uint64_t size);
+  BufferBase(const void* data, uint64_t size);
+
+  /** Returns the buffer data. */
+  void* nonconst_data() const;
+
+  /** Returns the buffer data as a pointer to constant. */
+  const void* const_data() const;
+
+  /** Returns the buffer at the current read offset */
+  void* nonconst_unread_data() const;
+
+  /** Returns the buffer at the current read offset as a pointer to constant */
+  const void* const_unread_data() const;
```

Comment:
not really 'read' offset for BufferBase is it, are these ('..._unread...') mnemonically appropriate for BufferBase at all?

Reply:
Changed.